crates/wysiwyg/src/dom/nodes/mention_node.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::composer_model::example_format::SelectionWriter;
+use crate::dom::dom_handle::DomHandle;
+use crate::dom::to_html::{ToHtml, ToHtmlExt, ToHtmlState};
+use crate::dom::to_markdown::{MarkdownError, MarkdownOptions, ToMarkdown};
+use crate::dom::to_plain_text::ToPlainText;
+use crate::dom::to_raw_text::ToRawText;
+use crate::dom::to_tree::ToTree;
+use crate::dom::unicode_string::{UnicodeStrExt, UnicodeStringExt};
+use crate::dom::UnicodeString;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MentionNode<S>
+where
+    S: UnicodeString,
+{
+    kind: MentionNodeKind<S>,
+    attributes: Vec<(S, S)>,
+    handle: DomHandle,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MentionNodeKind<S>
+where
+    S: UnicodeString,
+{
+    MatrixUrl { display_text: S, url: S },
+    AtRoom,
+}
+
+impl<S> MentionNode<S>
+where
+    S: UnicodeString,
+{
+    /// Create a new MentionNode
+    ///
+    /// NOTE: Its handle() will be unset until you call set_handle() or
+    /// append() it to another node.
+    pub fn new(url: S, display_text: S, attributes: Vec<(S, S)>) -> Self {
+        let handle = DomHandle::new_unset();
+
+        Self {
+            kind: MentionNodeKind::MatrixUrl { display_text, url },
+            attributes,
+            handle,
+        }
+    }
+
+    pub fn new_at_room(attributes: Vec<(S, S)>) -> Self {
+        let handle = DomHandle::new_unset();
+
+        Self {
+            kind: MentionNodeKind::AtRoom,
+            attributes,
+            handle,
+        }
+    }
+
+    pub fn name(&self) -> S {
+        S::from("mention")
+    }
+
+    pub fn display_text(&self) -> S {
+        match self.kind() {
+            MentionNodeKind::MatrixUrl { display_text, .. } => {
+                display_text.clone()
+            }
+            MentionNodeKind::AtRoom => S::from("@room"),
```

Comment:
Wonder if we should worry about localisation on that one, or just suppose the client will translate these back and forth when using something else than `@room`

Reply:
Good question about localisation. From what I found, all the documentation around `@room` mentions point to an an exact match on the string `"@room"` and no mention of localisation, so it would seem safe to do this ignore it